### DISCUSSION TITLE
Do mistral don't support system prompt?

### DISCUSSION THREAD

**Discussion by TikaToka** [ggml-org/llama.cpp#3932] [Q&A]
I am trying with interactive server, but it looks like it does not work

**Comment by cmp-nct** [comment]
Is there any LLM that supports a system prompt right now ? I don't think so.
Some come with one but they ignore changes to it mostly. In all LLMs I have tested you'll be better in doing the "system" prompt in the normal way of addressing it. that includes chatGPT/GPT4 which offer a system prompt that basically only ruins the output if you use it.

**Reply by KerfuffleV2** [reply]
Eric Hartford's models support a system prompt (actually, I believe any model that uses ChatML style should).
1. https://huggingface.co/TheBloke/dolphin-2.2.1-mistral-7B-GGUF
2. https://huggingface.co/TheBloke/Dolphin-2.1-70B-GGUF

I can't really say too much about exactly how much effect the system prompt has, because I haven't done much experimentation between using it or not but it definitely doesn't ruin the output.

The 7B is quite good for its size (I'm not really sure there's anything better). The 70B is the best I've tried in terms of how intelligent it seems.

**Reply by cmp-nct** [reply]
I'll give Dolphin a try.
By "ruin" I meant damage the model response.
For example, when I used GPT4 (a lot) for development tasks, the easiest approach would be a system prompt. To guide it into exactly the right type of response. The results were underwhelming, the model became significantly more stupid, it was not useable that way.
When looking at llava the system prompt barely plays a role, probably the same for all Vicuna type models.

The reason behind this, in my opinion, is that during the instruction/chat fine-tune the models are not tuned on a system prompt but impregnated with the same prompt over and over again + instruction/chat.
So if you deviate from the original system prompt it's a bit like removing the special tokens like ASSISTANT: or USER: HUMAN: etc.
The model might respond anyway but it loses a lot of the fine tuning.

**Reply by KerfuffleV2** [reply]
> By "ruin" I meant damage the model response.

Yes, I understood what you meant. "Decreased the quality of the model's responses in a noticeable way" in other words

> The reason behind this, in my opinion, is that during the instruction/chat fine-tune the models are not tuned on a system prompt but impregnated with the same prompt over and over again + instruction/chat.

Those Dolphin models aren't fine tuned for ChatML on top of some existing instruct model, they're trained on a base model. In the case of the 7B, it's on top of the base Mistral model. In the case of the 70B it's derived from StellarBright which is derived from the base LLaMA2 70B non-instruct model.

**Reply by KerfuffleV2** [reply]
You may or may not be right about the system prompt being helpful (in general, or for Hellaswag-type questions in specific) but I don't think _this_ test really proves anything.

> system prompt "You are Dolphin2.1, an AI assistant" HELLASWAG 7573 2469 10042 2 .754

The point of a system prompt is to steer the model toward answering in a certain way. What does "*You are Dolphin2.1, an AI assistant*" do? Was the model even trained to know what "Dolphin2.1" is specifically? (I'd guess not.) Does being told it's an "AI assistant" do anything to make it answer in a way that will increase its chances of answering Hellaswag questions better?

I think an appropriate system prompt for that would be something about how it's expert on multiple choices questions, always takes its time and thinks carefully, knows a lot of trivia about various stuff (what's a waist trainer? I don't even know), that kind of thing. Something that's going to prepare it with an approach that will make it more likely to produce the kind of output you're looking for (in this case, answering the multiple choice Hellaswag questions more accurately).


Also, like all prompting, it's a process and may take refining to get the best results.

**Reply by KerfuffleV2** [reply]
> I intentionally did not want to steer it in any direction with its responses, just informing it its name and the fact that it was an AI assistant.

I don't understand what you were testing or what the results are supposed to signify then. As I understand it, the point of the system prompt is to make the model do something.

It definitely doesn't surprise me that a system prompt with no purpose (and that in fact actually probably confuses the model) makes it perform worse.

> also blanking out the system prompt had a similar significant negative impact, so the only conclusion is that the built in system prompt, whatever it is, is helping the model perform better. 

It doesn't really work that way. There isn't a system prompt it will "assume" if you just leave the system prompt blank. Just for example, the Alpaca models that had "*Below is an instruction that describes a task. Write a response that appropriately completes the request.*" as the "system prompt" (which wasn't meant to be configurable) still required writing exactly that. You could skip it and move directly to prompting the model, but it was generally recommended not to.

So again, it doesn't really surprise me that a system prompt that doesn't do _anything_ to prepare the model for its role is going to perform worse than nothing on a model that was trained to use the system prompt. I mean, during training involving the system prompt, there would have been something actually meaningful in there.

**Reply by TikaToka** [reply]
@cmp-nct  @alt8git I agree that default system prompt is optimized value for chat purpose, but I am utilizing the LLMs to do another job (not just chatting, I need specific format of output to define pernamently), so utilizing system prompt is needed for me :(

@KerfuffleV2 thanks! i'll try dolphin version ;)

**Reply by KerfuffleV2** [reply]
@alt8git 

> I assumed the guy that created the model used that system prompt for a reason.

You changed it though, and also "Dolphin2.1" with the words run together like that could potentially confuse the model even if it understood "Dolphin".

Anyway, being told it's an AI assistant could make it function better for assistant type stuff. A lot of people would their assistant to be personable, for example. This isn't something that would help with the Hellaswag test though, and in fact could be harmful.

> thus confirming that either specifying or erasing system prompt with a blank can degrade model performance.

You really don't have to prove that and you definitely don't have to convince _me_. It absolutely is possible.

This conversation was in the context of the other person saying using a system prompt basically always noticeably decreases the quality of responses. I'm actually not even arguing the opposite of that: that there are _definitely_ cases where using the system prompt is a benefit. I'm just saying we haven't proven that.

**Reply by TikaToka** [reply]
@KerfuffleV2 @alt8git  I tried to use Dolphin, but there's an segmentation fault like
```
{"timestamp":1699425443,"level":"INFO","function":"main","line":2492,"message":"HTTP server listening","hostname":"127.0.0.1","port":9090}
all slots are idle and system prompt is empty, clear the KV cache
slot 0 is processing [task id: 0]
slot 0 : kv cache rm - [0, end)

print_timings: prompt eval time =     634.08 ms /   289 tokens (    2.19 ms per token,   455.78 tokens per second)
print_timings:        eval time =    3443.55 ms /   154 runs   (   22.36 ms per token,    44.72 tokens per second)
print_timings:       total time =    4077.64 ms
{"timestamp":1699425447,"level":"INFO","function":"log_server_request","line":2156,"message":"request","remote_addr":"127.0.0.1","remote_port":44238,"status":200,"method":"POST","path":"/completion","params":{}}
slot 0 released (444 tokens in cache)
updating system prompt
GGML_ASSERT: llama.cpp:5867: n_tokens <= n_batch
dolphin1.sh: line 2: 658919 Aborted                 (core dumped) CUDA_VISIBLE_DEVICES=6 ./server -m models/dolphin-2.2.1-mistral-7B-GGUF/dolphin-2.2.1-mistral-7b.Q5_K_M.gguf -c 2048 --port 9090 -ngl 35
```

I ran with
```
CUDA_VISIBLE_DEVICES=6 ./server -m models/dolphin-2.2.1-mistral-7B-GGUF/dolphin-2.2.1-mistral-7b.Q5_K_M.gguf -c 2048 --port 9090  -ngl 35
```

Do you know anything related to this error? before posting new issue or discussion, I wanted to ask you as I thought you might know this

**Reply by slaren** [reply]
The `n_tokens <= n_batch` assert fails when doing exactly that, an evaluation with more tokens than the batch size configured for the context. The evaluation needs to be split into multiple batches in that case, and if the server example doesn't do that, then that's probably a bug in the server example. However, we should probably change `llama_decode` to do that automatically for ease of use.
